[Android][image-picker] Rewrite module to Sweet API

bbarthec opened the pull request:
# I'm going to split this PR as it's grown up in side considerably

# Why

Resolves [ENG-3071](https://linear.app/expo/issue/ENG-3071/audit-expo-image-picker-android-rewrite-to-sweet-api)

# How

- Refactored `ImagePickerOptions.kt` to use auto-conversion features of Sweet API
- Refactored `ImagePickerModule.kt` to follow new `Module`'s syntax
- Removed unused `api "androidx.legacy:legacy-support-v4:1.0.0"`
- Added `implementation "androidx.core:core-ktx:1.7.0"` to be able to use `bundleOf` shorthand

# Test Plan

TBD

# TODO

- [ ] Test on actual devices all flows/configurations
    - [ ] picking image
        - [ ] image library
        - [ ] camera
     - [ ] picking video
       - [ ] image library
       - [ ] camera
     - [ ] additional params
         - [ ] base64
         - [ ] exif
     - [ ] multi-selection
     - [ ] picking all types of media
     - [ ] editing
     - [ ] check if `Activity` destruction scenario is working 
- [ ] Ensure the API hasn't changed
- [ ] Follow-up on each `TODO`

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

bbarthec committed:
Initial refactor into SweetAPI

Refactored the ImagePickerModule and ImagePickerOptions classes to follow new Sweet API syntax

barthap reviewed (commented):
Not an actual review, just first impressions 😁

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,63 @@
+  VIDEOS("Videos"),
+  ALL("All");
+
+  internal fun toIntent(): Intent {
+    return Intent().also {
```
```suggestion
  internal fun toIntent() = Intent().also {
```
Up to you, but IMO it looks better.

lukmccall reviewed (commented):
Looks goog ;)

tsapeta commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,63 @@
+import expo.modules.kotlin.records.Record
+
+internal class ImagePickerOptions: Record {
+  /**
+   * TODO(@bbarthec): restore preventing numbers from outside valid range
+   */
+  @Field
+  @FloatRange(from = 0.0, to = 1.0, fromInclusive = true, toInclusive = true)
+  var quality: Double = 0.2
```
Created linear task for this https://linear.app/expo/issue/ENG-4123/more-customized-arguments-validation

bbarthec committed:
[image-picker][Android] Delete `ImagePickerPackage.kt` as it's no longer needed

lukmccall commented on packages/expo-cellular/android/src/main/java/expo/modules/cellular/CellularModule.kt:
```diff
@@ -61,7 +61,7 @@ class CellularModule : Module() {
 
   private val context
-    get() = requireNotNull(appContext.reactContext)
+    get() = requireNotNull(appContext.reactContext, )
```
```suggestion
    get() = requireNotNull(appContext.reactContext)
```
❓

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,76 @@
-      val isAllowsEditing = options[ImagePickerConstants.OPTION_ALLOWS_EDITING] as? Boolean ?: false
+internal class ImagePickerOptions: Record {
+  @Field
+  @FloatRange(from = 0.0, to = 1.0)
```
```suggestion
  @DoubleRange(from = 0.0, to = 1.0)
```

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,76 @@
+   * TODO(@bbarthec): undocumented
+   */
+  @Field
+  @IntRange(from = 0)
```
@IntRange annotation isn't supported on Pair<Int, Int> type - it will lead to crash right now

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/exporters/CompressionImageExporter.kt:
```diff
@@ -12,9 +13,14 @@ import java.io.IOException
 class CompressionImageExporter(
   private val mImageLoader: ImageLoaderInterface,
-  private val mQuality: Int,
+  @FloatRange(from = 0.0, to = 1.0, fromInclusive = true, toInclusive = true)
```
```suggestion
  @DoubleRange(from = 0.0, to = 1.0, fromInclusive = true, toInclusive = true)
```

lukmccall replied:
You will lose accuracy here. I guess it doesn't matter in that case 🤷

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/assertions/Assertions.kt:
```diff
@@ -0,0 +1,25 @@
+      "not smaller than $lowerBound"
+    else
+      "not greater than $upperBound"
+  }")
```
```suggestion
  }")

```

bbarthec committed:
[emc][Android] Add `ActivityResultsRegistry`

`ReactNative` apps that uses `emc` cannot hook propoerly into `Activity` lifecycle and thus we cannot cannot directly use all fancy stuff from `AndroidX`.
I'm proposing similar solution for registering for results coming from 3rd party Activities as AndroidX's https://developer.android.com/training/basics/intents/result.

- Create `AppContextActivityResultCallback` that mirrors https://developer.android.com/reference/androidx/activity/result/ActivityResultCallback and has additional parameter `launchingActivityHasBeenKilled`
- Create `AppContextActivityResultCaller` that mirrors https://developer.android.com/reference/androidx/activity/result/ActivityResultCaller to accepts `AppContextActivityResultCallback`
- Create `AppContextActivityResultRegistry` that mirros https://developer.android.com/reference/androidx/activity/result/ActivityResultRegistry, but does not throw upon too late initialization. The original class forces registration in `onCreate` lifecycle method and when using `ReactNative` we are not able to hook into that method properly. Moreover the original class is coupled with Activity's lifecycle and we need something that would outlive Activity (and is hooked into `AppContext` lifecycle)
- Create `ActivityResultsManager` that is encapsulating all `Activity`-related-for-result-registration outside `AppContext`

bbarthec committed:
[amc][Android] Expose `registerForActivityResult` and `launchForActivityResult` via `AppContext`

bbarthec committed:
[emc][Android] Create `AppCompatActivityProvider`

The hosting [Activity] is always `AppCompatActivity` instance, becasue `ReactActivity` is subclassing this AndroidX's class.
We cannot expose `ReactActivity` instead as that would force us to depend on `react-native` directly and we don't want it for now in `expo-modules-core`.

bbarthec committed:
[emc][Android] Create `AppContextProvider`

Some modules might want to access `AppContext` and this interface would enable them to do so without retaining the `AppContext` exlicitly.
It would also allow us to pass the `AppContext` to the other classes the module might use/depend on.

bbarthec committed:
[emc][Android] Create `ContextProvider`

Some modules might want to access `Context` and this interface would allow it.
It also allows passing `Context` to the other classes without referencing it explicitly.

bbarthec committed:
[emc][Android] Create `CoroutineScopeProvider`

This interface allows us to acces `CoroutineScope` in the modules.

bbarthec committed:
[emc][Android] Add missing `ActivityResultsManager` and missing `coroutines` imports

bbarthec committed:
[emc][Android] Change `Context` to more specific `ReactContext`

We're already referencing `ReactApplicationContext` from `AppContext`, so there's no reason to generalize other `React`-related stuff.

bbarthec committed:
[emc][Android] Implement `AppCompatActivityProvider` in `AppContext`

bbarthec committed:
[emc][Android] Deprecate `activityProvider` in favour of more specific `currentAppCompatActivity` in `AppContext`

bbarthec committed:
[emc][Android] Add `androidx.appcompat:appcompat:1.4.1` and `androidx.activity:activity-ktx:1.4.0` dependencies

In order to take advantage of newer AndroidX features we need to enforce higher version of `appcompat` then ReactNative.
ReactNative is bumping this library, but it's still to be released.
Moreover I've added direct dependecy on `activity-ktx` to have even more nice featureas available.
I've also bumped `androidx.annotation:annotation@1.2.0 ➡️ 1.3.0`

bbarthec committed:
[emc][Android] Deprecate `ActivityProvider` altogether

bbarthec committed:
[emc][Android] Make `Module` implement `AppContextProvider`, `ContextProvider` and `CoroutineScopeProvider`

Each module should have access to `AppContext`, `Context` or `CoroutineScope` by default.

bbarthec committed:
[clipboard][eas-client] Adapt new `ContextProvider`

bbarthec committed:
[image-picker][Android] Upgrade `build.gradle`

Add dependencies
- `androidx.activity:activity-ktx:1.4.0`

Bump dependencies:
- `com.github.CanHub:Android-Image-Cropper@1.1.1 ➡️ 3.3.5`
- `org.jetbrains.kotlinx:kotlinx-coroutines-android@1.4.1 ➡️ 1.6.0`
- `org.jetbrains.kotlinx:kotlinx-coroutines-core@1.4.3 ➡️ 1.6.0`

Remove dependencies:
- `commons-codec:commons-codec`
- `commons-io:commons-io`

bbarthec committed:
[image-picker][Android] Temporarily bump `compileSdkVersion` and `targetSdkVersion` to 32

bbarthec committed:
[image-picker][Android] Move `ImagePickerFIleProvider` to `fileproviders` directory

bbarthec committed:
[image-picker][Android] Cleanup `ImagePickerConstants` by removing redundant code

bbarthec committed:
[image-picker][Android] Create unified `ImagePickerExceptions`

bbarthec committed:
[image-picker][Android] Create `Options` and `Response` for JS <> Android communication

- Reanmed `forceAspect` to `aspect` to match docs
- Added a few experimental validations

bbarthec committed:
[emc][Android] Invoke `setter` method for proppoerty annotated with `@Field`

bbarthec committed:
[emc][Android] Add generic `ModuleNotFoundException`

bbarthec committed:
[emc][Android] Make `DecoratedException` open and `Throwable` as `cause` (was `CodedException`)

bbarthec committed:
[emc][Android] Catch and properly handle `InvocationTargetException` (coming from coroutines) in our exception decoration pipeline

bbarthec committed:
[emc][Android][Proposal] Propose alterantive validation mechanism for `@Field`s

bbarthec committed:
[emc][Android] Force olf permissions interface accept new Sweet API Promise

bbarthec committed:
[image-picker][Android] Create `CameraContract`, `ImageLibraryContract` and `CropImageContract`

Create contracts to leverage new `registerForActivityResult` API that is exposed by `AppContext`

bbarthec committed:
[image-picker][Android] Refactor file exporters

Refactor and simplify media assets exporters that export picked asset to a file and provide some additional methods to acces more info about picked media.

bbarthec committed:
[image-picker][Android] Refactor file providers

Possibly to removed altogether.

bbarthec committed:
[image-picker][Android] Remove old `Image/VideoResultTask`s

bbarthec committed:
[image-picker][Android] Remove `PendingPromise` and `PickerResultsStore` as these are replaced by `registerForActivityResult` mechanism

bbarthec committed:
[image-picker][Android] Create `MediaHandler`

`MediaHandler` is a class that has to convert raw `Uri` (picking result) into the actual result that might be sent to the JS.

bbarthec committed:
[image-picker][Android] Refactor utils

bbarthec committed:
[image-picker][Android] Refactor `IamgePickerModule`

I've delegated as much as possible to other classes.
This class only registers the exported methods and handles very high level flow of each method.

It's using following mechanisms:
- `ContextProvider`, `AppContextProvider` that are forwarded to `MediaHandler`
- `launchForActivityResult` to propoerly call 3rd party Activities and monitors they possible destruction
- the only state that is actually stored in this class narrows down to  `PendingMediaPickingResult` that might come in handy whenever the hosting Activity is destroyed

bbarthec committed:
[image-picker] Regsiter module as Sweet API compatible

bbarthec committed:
[expoview] Register new `image-picker` module

bbarthec committed:
[bare-expo][Android] Solution for `rn-screens` problem https://github.com/software-mansion/react-native-screens/issues/17#issuecomment-424704067

bbarthec committed:
[ncl] Handle error result gracefully in `FunctionDemo` component

`FunctionComponent` can now display errors/rejections.
It also renders/logs some meaningful error message now.

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,119 @@
+  @Field
+  @FloatRange(from = 0.0, to = 1.0)
+  var quality: Double = 0.2
+    set(value) {
+      // TODO(@bbarthec): possibly we don't need that
+      assertValueInRange(value, lowerBound = 0.0, upperBound = 1.0)
+      field = value
+    }
```
```suggestion
```
You don't need that as long as you're using the validators from e-m-c

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,119 @@
+  var mediaTypes: MediaTypes = MediaTypes.IMAGES
+
+  @Field
+  @IntRange(from = 0)
+  var videoMaxDuration: Int = 0
+    set(value) {
+      assertValueGreaterOrEqual(value, 0)
+      field = value
+    }
```
```suggestion
  @IntRange(from = 0)
  var videoMaxDuration: Int = 0
```
same here

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerUtils.kt:
```diff
@@ -1,92 +1,149 @@
+    Bitmap.CompressFormat.PNG
+  }
+  else -> {
+    if (!this.contains("jpeg")) {
```
Similar to the comment above. 

I would choose a separate case for `contains("jpeg")` and handle that unknown type in the `else` block.

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerUtils.kt:
```diff
@@ -1,92 +1,149 @@
+}
+
+/**
+ * The same as [CoroutineScope.launch] but with built-in exception handler
+ * that relays caught exceptions to [Promise.reject] ensuring that these are wrapped in [CodedException].
+ *
+ * TODO(@bbarthec): inspect whether not to add this to the [AsyncFunction] definition
```
It was already built into the `e-m-c`.

If you want to transform the function's body into the `coroutine`, you need to use a construction like this:
```
AsyncFunction("<name>")  Coroutine { <body> }
```
It should also handle all uncaught exceptions.

lukmccall replied:
And also, if you want to dispatch some coroutines, you can use a `coroutineScope` from the module class or if you're in the `AsyncFunction ... Coroutine` block, you will have access to the function scope there.

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/MediaHandler.kt:
```diff
@@ -0,0 +1,85 @@
+      exif = exif,
+    )
+  }
+
```
```suggestion
```

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/MediaHandler.kt:
```diff
@@ -0,0 +1,85 @@
+    ).generateFile()
+
+    val exportedImage = exporter.exportAsync(sourceUri, outputFile)
+    val base64 = options.base64.takeIf { it }
+      ?.let { exportedImage.data() }
+      ?.let { Base64.encodeToString(it.toByteArray(), Base64.NO_WRAP) }
```
An interesting way of writing an `if` statement  😅

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/crop/ImageLibraryContract.kt:
```diff
@@ -0,0 +1,60 @@
+import androidx.activity.result.contract.ActivityResultContract
+import expo.modules.imagepicker.MediaType
+
+
```
```suggestion
```

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/exporters/CompressionImageExporter.kt:
```diff
@@ -2,68 +2,70 @@ package expo.modules.imagepicker.exporters
-  private val mBase64: Boolean
+  private val appContextProvider: AppContextProvider,
+  private val contextProvider: ContextProvider,
+  @FloatRange(from = 0.0, to = 1.0)
```
```suggestion
  @FloatRange(from = 0.0, to = 1.0)
```
Use the same validator from e-m-c

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/AppContextActivityResultCaller.kt:
```diff
@@ -0,0 +1,27 @@
+    contract: ActivityResultContract<I, O>,
+    callback: AppContextActivityResultCallback<O>
+  ): ActivityResultLauncher<I>
+}
```
```suggestion
}

```

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/AppContextActivityResultCallback.kt:
```diff
@@ -0,0 +1,14 @@
+   * has been recreated from scratch as well and all previous state is invalid.
+   */
+  fun onActivityResult(result: O, launchingActivityHasBeenKilled: Boolean)
+}
```
```suggestion
}

```

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,119 @@
+  VIDEOS("Videos"),
+  ALL("All");
+
+  internal fun toMimeType(): String {
+    return when (this) {
+      IMAGES -> ImageAllMimeType
+      VIDEOS -> VideoAllMimeType
+      ALL -> AllMimeType
+    }
+  }
```
I prefer this syntax, but up to you
```suggestion
  internal fun toMimeType(): String = when (this) {
    IMAGES -> ImageAllMimeType
    VIDEOS -> VideoAllMimeType
    ALL -> AllMimeType
  }
```

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,119 @@
+    }
+  }
+
+  internal fun toMultipleMimeTypes(): Array<String>? {
+    return when (this) {
+      ALL -> arrayOf(
+        ImageAllMimeType,
+        VideoAllMimeType
+      )
+      else -> null
+    }
+  }
```
```suggestion
  internal fun toMultipleMimeTypes(): Array<String>? = when (this) {
    ALL -> arrayOf(
      ImageAllMimeType,
      VideoAllMimeType
    )
    else -> null
  }
```

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,119 @@
-        promise.reject(ImagePickerConstants.ERR_INVALID_OPTION, "videoMaxDuration must be a non-negative integer")
-        return null
-      } ?: 0
+  internal fun toFileExtension(): String {
+    return when(this) {
+      VIDEOS -> ".mp4"
+      else -> ".jpeg"
+    }
+  }
```
```suggestion
  internal fun toFileExtension(): String = when(this) {
    VIDEOS -> ".mp4"
    else -> ".jpeg"
  }
```

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerOptions.kt:
```diff
@@ -1,58 +1,119 @@
     }
   }
+
+  internal companion object {
+    const val ImageAllMimeType = "image/*"
+    const val VideoAllMimeType = "video/*"
+    const val AllMimeType = "*/*"
+  }
```
- I think we keep `const val` names uppercase: `IMAGE_ALL_MIME_TYPE`
- Maybe we could create an enum for this? Something like:

```kt
internal enum class PickerMediaTypes(val mimeType: String) {
  IMAGES("image/*"),
  VIDEOS("video/*"),
  ALL("*/*")
}
```

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerResponse.kt:
```diff
@@ -0,0 +1,43 @@
+  ): ImagePickerMediaResponse(MediaType.VIDEO, uri, width, height)
+}
+
+enum class MediaType(val value: String) {
```
If we created the enum for mime type options above (see my previous comment), we could rename this one for sth like `ResultMediaType`

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/crop/CropImageContract.kt:
```diff
@@ -0,0 +1,98 @@
+  private val options: ImagePickerOptions,
+  private val pickingSource: PickingSource
+): MediaPickerContract() {
+//  TODO(@bbarthec): do we need it anymore?
+//  val exifDataHandler = ExifDataHandler(sourceUri)
+//
+//  class ExifDataHandler(private val uri: Uri) {
```
I think so. IIRC this is needed to copy EXIF tags from original to cropped image, when user set `allowsEditing: true`

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/crop/ImageLibraryContract.kt:
```diff
@@ -0,0 +1,60 @@
+    data = intent.takeIf { resultCode == Activity.RESULT_OK }?.getClipDataUris() ?: emptyList()
+  )
+
+  internal companion object {
+    internal fun Intent.getClipDataUris(): List<Pair<MediaType, Uri>> {
```
- I'd extract this function from companion object to global file scope
- Isn't `List<Pair<X, Y>>` just a `Map<X, Y>` 😅 If you need list of pairs, any map has a `.toList()` function for that

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/exporters/ImageExporter.kt:
```diff
@@ -1,21 +1,75 @@
+            }
+          }
+        // Explicitly get latitude, longitude, altitude with their specific accessor functions.
+        exifInterface.latLong?.let { latLong ->
+          putDouble(ExifInterface.TAG_GPS_LATITUDE, latLong[0])
+          putDouble(ExifInterface.TAG_GPS_LONGITUDE, latLong[1])
```
```suggestion
        exifInterface.latLong?.let { (lat, lng) ->
          putDouble(ExifInterface.TAG_GPS_LATITUDE, lat)
          putDouble(ExifInterface.TAG_GPS_LONGITUDE, lng)
```

barthap reviewed (commented):
One iteration 😉

bbarthec commented:
Almost all of the work is now delivered in separate PRs 😉
